File: rb/word_generator.rb
require 'concord'

DICTIONARY = ['foo', 'bar', 'baz', 'qux']

def log(str)
  $stderr.puts str
  $stderr.flush
end

def time_millis()
  (Time.now.to_f * 1000.0).to_i
end

class WordGenerator
  # init takes one argument:
  # - context: the context object used to interact with the framework
  def init(context)
    future = time_millis
    log "Initialized Computation"
    context.set_timer('default', time_millis)
  end

  # process_timer takes three arguments:
  # - context: the context object used to interact with the framework
  # - key: the name of the timer that's being triggered
  # - time: the time the timer was scheduled to trigger at
  def process_timer(context, key, time)
    (0..1024).each do |i|
      context.produce_record('words', DICTIONARY.sample, '')
    end
    context.set_timer(key, time_millis + 5000)
  end

  # metadata takes no arguments, but expects a return value of a `Metadata`
  # object. check out the ruby client api documentation for more information.
  def metadata
    log "Metadata called"
    Concord::Metadata.new(name: 'word-generator', ostreams: ['words'])
  end
end

Concord::Computation.serve(WordGenerator.new)

File: cpp/word_generator.cc
#include <chrono>
#include <cstdint>
#include <vector>
#include <random>
#include <concord/glog_init.hpp>
#include <concord/Computation.hpp>
#include <concord/time_utils.hpp>

static const std::vector<std::string> kOurFamousLines = {
  "How", "do", "I", "love", "thee", "Let", "me", "count"};

class WordGenerator final : public bolt::Computation {
  public:
  using CtxPtr = bolt::Computation::CtxPtr;

  virtual void init(CtxPtr ctx) override {
    LOG(INFO) << "Initializing word generator [cpp]";
    ctx->setTimer("loop", bolt::timeNowMilli());
  }

  virtual void destroy() override {
    LOG(INFO) << "Destructing word generator [cpp]";
  }

  virtual void
  processTimer(CtxPtr ctx, const std::string &key, int64_t time) override {
    for(auto i = 0u; i < 10000; ++i) {
      std::string bin = kOurFamousLines[i % kOurFamousLines.size()];
      ctx->produceRecord("words", bin, "");
    }
    ctx->setTimer(key, bolt::timeNowMilli());
  }

  virtual bolt::Metadata metadata() override {
    bolt::Metadata m;
    m.name = "word-generator";
    m.ostreams.insert("words");
    return m;
  }

  virtual void processRecord(CtxPtr ctx, bolt::FrameworkRecord &&r) override {}
};

int main(int argc, char *argv[]) {
  // int i = 0; while(!i){sleep(1);} // gdb hack
  bolt::logging::glog_init(argv[0]);
  bolt::client::serveComputation(std::make_shared<WordGenerator>(), argc, argv);
  return 0;
}
